PowerEditor/src/WinControls/ProjectPanel/ProjectPanel.cpp

```diff
@@ -482,18 +603,105 @@ bool ProjectPanel::buildTreeFrom(TiXmlNode *projectRoot, HTREEITEM hParentItem)
 					return false;
 			}
 		}
+		if (lstrcmp(TEXT("Directory"), v) == 0)
+		{
+
+			const TCHAR *strLabel = (childNode->ToElement())->Attribute(TEXT("userLabel"));
+			generic_string newFolderLabel;
+			
+			const TCHAR *strValue = (childNode->ToElement())->Attribute(TEXT("name"));
+			generic_string fullPath = getAbsoluteFilePath(strValue);
+
+			std::vector<generic_string> filters;
+			const TCHAR *strFilters = (childNode->ToElement())->Attribute(TEXT("filters"));
+			if (strFilters)
+			{
+				generic_string filter(strFilters);
+				filters = stringSplit(filter, TEXT(";"), true);
+			}
+
+			if (strLabel)
+			{
+				newFolderLabel = strLabel;
+			}
+			else
+			{
+
+				newFolderLabel = buildDirectoryName(fullPath, filters);
+			}
+
+
+			HTREEITEM hNewDirectory = addDirectory(hParentItem, newFolderLabel.c_str(), true, fullPath.c_str(), &filters);
+			if (strLabel)
+			{
+				TVITEM tvItem;
+				tvItem.mask = TVIF_PARAM;
+				tvItem.hItem = hNewDirectory;
+				::SendMessage(_treeView.getHSelf(), TVM_GETITEM, 0,reinterpret_cast<LPARAM>(&tvItem));
+				ProjectPanelData& data = *reinterpret_cast<ProjectPanelData*>(tvItem.lParam);
+				data._userLabel = strLabel;
+			}
+
+		}
 		else if (lstrcmp(TEXT("File"), v) == 0)
 		{
 			const TCHAR *strValue = (childNode->ToElement())->Attribute(TEXT("name"));
 			generic_string fullPath = getAbsoluteFilePath(strValue);
 			TCHAR *strValueLabel = ::PathFindFileName(strValue);
 			int iImage = ::PathFileExists(fullPath.c_str())?INDEX_LEAF:INDEX_LEAF_INVALID;
-			_treeView.addItem(strValueLabel, hParentItem, iImage, fullPath.c_str());
+			_treeView.addItem(strValueLabel, hParentItem, iImage, new ProjectPanelData(_directoryWatcher, strValueLabel, fullPath.c_str(), nodeType_file));
 		}
 	}
 	return true;
 }
 
+void ProjectPanel::rebuildDirectoryTree(HTREEITEM hParentItem, const ProjectPanelData& data)

```

Comment:
[Here's a neat trick](http://ariccio.com/2015/04/02/preventing-bugs-and-improving-code-quality-with-microsoft-sal-part-2-custom-preconditions-for-structs-objects/) to help detect (via static analysis) conditions that might lead to the `assert(0);` at line 692:

```
_Pre_satisfies_( ( data._nodeType == nodeType_baseDir ) || ( data._nodeType == nodeType_dir ) )
void ProjectPanel::rebuildDirectoryTree(HTREEITEM hParentItem, const ProjectPanelData& data)
```